backend/app/Savor22b/Action/PlaceUserHouseAction.cs

```diff
@@ -39,103 +34,161 @@ public PlaceUserHouseAction(int villageID, int targetX, int targetY)
             [nameof(TargetY)] = TargetY.Serialize(),
         }.ToImmutableDictionary();
 
-    protected override void LoadPlainValueInternal(
-        IImmutableDictionary<string, IValue> plainValue)
-    {
-        VillageID = plainValue[nameof(VillageID)].ToInteger();
-        TargetX = plainValue[nameof(TargetX)].ToInteger();
-        TargetY = plainValue[nameof(TargetY)].ToInteger();
-    }
-
-    private bool isAbleToPlaceHouse(Village village, int targetX, int targetY)
+    private static Village GetVillage(int villageID)
     {
-        int halfWidth = (village.Width - 1) / 2;
-        int halfHeight = (village.Height - 1) / 2;
-
-        // halfWidth는 중심점을 기준으로 좌우 몇칸 설치가 가능한 지를 나타냄. 그러니 절댓값 targetX가 halfWidth보다 크면 설치 불가
-
-        if (Math.Abs(targetX) > halfWidth)
-        {
-            return false;
-        }
+        var village = CsvDataHelper.GetVillageByID(villageID);
 
-        if (Math.Abs(targetY) > halfHeight)
+        if (village == null)
         {
-            return false;
+            throw new InvalidVillageException("Invalid village ID");
         }
 
-        return true;
+        return village;
     }
 
-    public override IAccountStateDelta Execute(IActionContext ctx)
+    private static void ValidateReplaceUserHouse(
+        long currentBlock,
+        RelocationState? relocationState
+    )
     {
-        IAccountStateDelta states = ctx.PreviousStates;
-
-        var village = CsvDataHelper.GetVillageByID(VillageID);
-
-        if (village == null)
+        if (relocationState is not null && relocationState.IsRelocationInProgress(currentBlock))
         {
-            throw new InvalidVillageException("Invalid village ID");
+            throw new RelocationInProgressException("Relocation in progress");
         }
+    }
 
-        if (!isAbleToPlaceHouse(village, TargetX, TargetY))
+    private void ValidateValidateHouseTarget()
+    {
+        Village village = GetVillage(VillageID);
+
+        if (!village.AbleToPlaceHouse(TargetX, TargetY))
         {
             throw new InvalidVillageException("Invalid target position");
         }
+    }
 
-        GlobalUserHouseState globalUserHouseState =
-            states.GetState(Addresses.UserHouseDataAddress) is Bencodex.Types.Dictionary stateEncoded
-                ? new GlobalUserHouseState(stateEncoded)
-                : new GlobalUserHouseState();
-
+    private void ValidateCheckPlacedHouse(GlobalUserHouseState globalUserHouseState)
+    {
         string userHouseKey = globalUserHouseState.CreateKey(VillageID, TargetX, TargetY);
 
-        if (globalUserHouseState.UserHouse.ContainsKey(userHouseKey))
+        if (globalUserHouseState.CheckPlacedHouse(userHouseKey))
         {
             throw new HouseAlreadyPlacedException("House already placed");
         }
+    }
+
+    private void ValidatePlaceHouse(GlobalUserHouseState globalUserHouseState)
```

Comment:
```suggestion
    private void ValidateForPlaceHouse(GlobalUserHouseState globalUserHouseState)
```
연결이 자연스럽지가 않아서 for 들어가야될 것 같습니다(근데 영알못이라 틀릴 수도 있습니다)

Reply:
넹